### PR TITLE
Version 0.5 Release

### PR DESCRIPTION
**Pull request by rstoneback**
Implemented a new algorithm that produces first orthogonal geomagnetic basis vector system that works for non-Spherical planets as well as for multi-pole magnetic fields.

### CODE REVIEW

**Comment by rstoneback** [comment]
I'm planning on releasing this version next week. Pinging @aburrell and @jklenzing as an FYI. I'm certainly open to code review though given the size of the pull it is not expected.

File: pysatMagVect/tests/test_comparison.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        self.inst.yr = 2019
+        self.inst.doy = 365
+        self.inst.date = pysat.datetime(2019, 12, 31)
+        self.dview = dview
+        self.dc = dc
+
+        return
+
+
+    def test_apexpy_v_pymv(self):
```
Comment (aburrell) [inline pysatMagVect/tests/test_comparison.py:22]: Missing docstrings

File: pysatMagVect/tests/test_comparison.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+        # dot product of zonal and field aligned
+        dot_apex_zonal = apex_fa[0,:]*apex_zon[0,:] + apex_fa[1,:]*apex_zon[1,:] + apex_fa[2,:]*apex_zon[2,:]
+        dot_apex_mer = apex_fa[0,:]*apex_mer[0,:] + apex_fa[1,:]*apex_mer[1,:] + apex_fa[2,:]*apex_mer[2,:]
+
+        dotmagvect_zonal = ze*fe + zn*fn + zu*fu
+        dotmagvect_mer = me*fe + mn*fn + mu*fu
+
+
+        try:
```
Comment (aburrell) [inline pysatMagVect/tests/test_comparison.py:92]: Is making a figure the most efficient way to do a unit test?

File: pysatMagVect/_core.py [hunk 3]
```diff
@@ -109,10 +109,12 @@ def geodetic_to_ecef(latitude, longitude, altitude):
 
     return x, y, z
 
+import pysatMagVect.fortran_coords
+ecef_to_geodetic = pysatMagVect.fortran_coords.ecef_to_geodetic
 
-def ecef_to_geodetic(x, y, z, method=None):
+def python_ecef_to_geodetic(x, y, z, method=None):
```
Comment (aburrell) [inline pysatMagVect/_core.py:117]: Curious about this name change

File: pysatMagVect/_core.py [hunk 4]
```diff
@@ -960,331 +1389,345 @@ def closed_loop_edge_lengths_via_footpoint(glats, glons, alts, dates, direction,
...
+                                   max_steps=None, e_field_scaling_only=None,
+                                   edge_length=None, edge_steps=None,
+                                   **kwargs):
     """
     Calculates scalars for translating ion motions at position
     glat, glon, and alt, for date, to the footpoints of the field line
     as well as at the magnetic equator.
-    
+
     All inputs are assumed to be 1D arrays.
```
Comment (aburrell) [inline pysatMagVect/_core.py]: should be shorter

**Review by aburrell — COMMENTED** [review]
Some general notes and one question.

File: .travis.yml [hunk 5]
```diff
@@ -1,44 +1,44 @@
...
-  #coverage run --source=pysat setup.py test
-after_success:
-  coveralls
+language: python
+dist: xenial
+matrix:
+  include:
+    - python: 2.7
+    - python: 3.6
+    - python: 3.7
```
Comment (jklenzing) [inline .travis.yml:7]: Should we add python 3.8?

File: pysatMagVect/_core.py [hunk 6]
```diff
@@ -23,7 +23,7 @@
```
Comment (jklenzing) [inline pysatMagVect/_core.py:23]: 2 blank lines before functions

File: pysatMagVect/_core.py [hunk 7]
```diff
@@ -90,12 +90,12 @@ def geodetic_to_ecef(latitude, longitude, altitude):
...
         Geodetic Height (km) above WGS84 reference ellipsoid.
-        
+
     Returns
     -------
     x, y, z
         numpy arrays of x, y, z locations in km
-        
+
     """
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: should be one blank line

File: pysatMagVect/_core.py [hunk 8]
```diff
@@ -177,11 +179,11 @@ def ecef_to_geodetic(x, y, z, method=None):
 
 def enu_to_ecef_vector(east, north, up, glat, glong):
```
Comment (jklenzing) [inline pysatMagVect/_core.py:183]: 2 blank lines before functions, here and throughout

File: pysatMagVect/_core.py [hunk 9]
```diff
@@ -346,87 +348,103 @@ def field_line_trace(init, date, direction, height, steps=None,
...
+    if recursive_loop_count is None:
         recursive_loop_count = 0
-    #     
+    # number of times integration routine must output step location
     if steps is None:
         steps = np.arange(max_steps)
+    # ensure date is a float for IGRF call
     if not isinstance(date, float):
         # recast from datetime to float, as required by IGRF12 code
         doy = (date - datetime.datetime(date.year,1,1)).days
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: missing whitespace after ','

File: pysatMagVect/_core.py [hunk 10]
```diff
@@ -541,31 +569,27 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
     be = [];
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: semicolons?

File: pysatMagVect/_core.py [hunk 11]
```diff
@@ -541,31 +569,27 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
-        # trace = full_field_line(init, time, ref_height, step_size=step_size, 
-        #                                                 max_steps=max_steps,
-        #                                                 steps=steps)
-        trace_north = field_line_trace(init, time, 1., ref_height, steps=steps,
-                                        step_size=step_size, max_steps=max_steps)
-        trace_south = field_line_trace(init, time, -1., ref_height, steps=steps,
-                                        step_size=step_size, max_steps=max_steps)
+        trace = full_field_line(init, time, ref_height, step_size=step_size,
+                                                        max_steps=max_steps,
+                                                        steps=steps)
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: over-indentation

File: pysatMagVect/_core.py [hunk 12]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
+                'd_fa_z' : d_fa_z,
+                'e_zon_x' : e_zon_x,
+                'e_zon_y' : e_zon_y,
+                'e_zon_z' : e_zon_z,
+                'e_mer_x' : e_mer_x,
+                'e_mer_y' : e_mer_y,
+                'e_mer_z' : e_mer_z,
+                'e_fa_x' : e_fa_x,
+                'e_fa_y' : e_fa_y,
+                'e_fa_z' : e_fa_z,
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: whitespace before ':'

File: pysatMagVect/_core.py [hunk 13]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
+                    'loops' : loop_num,
+                    'vector_seed_type' : init_type,
+                    'diff_zonal_vec' : diff_z,
+                    'diff_mer_vec' : diff_m,
+                    'd_zon2_x' : d_zon2_x,
+                    'd_zon2_y' : d_zon2_y,
+                    'd_zon2_z' : d_zon2_z,
+                    'd_mer2_x' : d_mer2_x,
+                    'd_mer2_y' : d_mer2_y,
+                    'd_mer2_z' : d_mer2_z,
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: whitespace before ':'

File: pysatMagVect/_core.py [hunk 14]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
+            for key in tempd.keys():
+                outd[key] = tempd[key]
+        return zx, zy, zz, bx, by, bz, mx, my, mz, outd
+
+    # return unit vectors for magnetic drifts in ecef coordinates
+    return zx, zy, zz, bx, by, bz, mx, my, mz
+
+
+
+def step_along_mag_unit_vector(x, y, z, date, direction=None, num_steps=1.,
```
Comment (jklenzing) [inline pysatMagVect/_core.py:1180]: Too many blank lines

File: pysatMagVect/_core.py [hunk 15]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
-        y = y + step_size*uy[0]*scalar
-        z = z + step_size*uz[0]*scalar
-            
-    return np.array([x, y, z])
-   
-    
-def apex_location_info(glats, glons, alts, dates):
+        x = x + step_size*ux#*scalar
+        y = y + step_size*uy#*scalar
+        z = z + step_size*uz#*scalar
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: Are these commented bits from an old version?  Or inline comments?

File: pysatMagVect/_core.py [hunk 16]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
+        If True, footpoint locations returned as lat, long, alt.
+
+    Returns
+    -------
+    array(len(glats), 3), array(len(glats), 3)
+        Northern and Southern ECEF X,Y,Z locations
+
+    """
+
+   # use input location and convert to ECEF
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: indentation

File: pysatMagVect/_core.py [hunk 17]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
+    steps = np.arange(num_steps+1)
+    for ecef_x, ecef_y, ecef_z, date in zip(ecef_xs, ecef_ys, ecef_zs, dates):
+        yr, doy = pysat.utils.time.getyrdoy(date)
+        double_date = float(yr) + float(doy) / 366.
+
+        root[:] = (ecef_x, ecef_y, ecef_z)
+        trace_north = field_line_trace(root, double_date, 1., 120.,
+                                    steps=steps,
+                                    step_size=step_size,
+                                    max_steps=num_steps)
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: indentation doesn't match paren (same in southern function below

File: pysatMagVect/_core.py [hunk 18]
```diff
@@ -600,353 +624,758 @@ def calculate_mag_drift_unit_vectors_ecef(latitude, longitude, altitude, datetim
...
+
+def apex_edge_lengths_via_footpoint(*args, **kwargs):
+    raise DeprecationWarning('This method now called apex_distance_after_footpoint_step.')
+    apex_distance_after_footpoint_step(*args, **kwargs)
+
+def apex_distance_after_footpoint_step(glats, glons, alts, dates, direction,
+                                    vector_direction, step_size=None,
+                                    max_steps=None, steps=None,
+                                    edge_length=25., edge_steps=5,
+                                    ecef_input=False):
```
Comment (jklenzing) [inline pysatMagVect/_core.py]: indentation doesn't match paren

File: pysatMagVect/_core.py [hunk 19]
```diff
@@ -960,331 +1389,345 @@ def closed_loop_edge_lengths_via_footpoint(glats, glons, alts, dates, direction,
...
+    # get footpoint location information
+    north_ftpnt, south_ftpnt = footpoint_location_info(ecef_xs, ecef_ys, ecef_zs,
+                                                       dates, ecef_input=True)
+
+    # prepare output memory
+    out = {}
+
+    # D and E vectors at user supplied location
+    # good for mapping to magnetic equator
+    _, _, _, _, _, _, _, _, _, infod = calculate_mag_drift_unit_vectors_ecef(ecef_xs, ecef_ys, ecef_zs, dates,
```
Comment (jklenzing) [inline pysatMagVect/_core.py:1704]: This function is defined above and is called like this every time.  Could we reconfigure the returned variables to aid in clarity?

File: pysatMagVect/satellite.py [hunk 20]
```diff
@@ -264,7 +261,7 @@ def add_mag_drift_unit_vectors(inst, max_steps=40000, step_size=10.):
                                           'of the IVM also expressed in ECEF (sc_*hat_*).'),
```
Comment (jklenzing) [inline pysatMagVect/satellite.py]: indentation

File: pysatMagVect/satellite.py [hunk 21]
```diff
@@ -345,26 +342,26 @@ def add_mag_drift_unit_vectors(inst, max_steps=40000, step_size=10.):
...
     None
         Instrument object modified in place
-    
+
     """
-    
+
     inst['iv_zon'] = {'data':inst['unit_zon_x'] * inst['iv_x'] + inst['unit_zon_y']*inst['iv_y'] + inst['unit_zon_z']*inst['iv_z'],
                       'units':'m/s',
                       'long_name':'Zonal ion velocity',
```
Comment (jklenzing) [inline pysatMagVect/satellite.py]: whitespace after ':'

File: pysatMagVect/tests/test_apex.py [hunk 22]
```diff
@@ -0,0 +1,716 @@
...
+
+        dx = pds.DataFrame(dx)
+        dy = pds.DataFrame(dy)
+        dz = pds.DataFrame(dz)
+        dh = pds.DataFrame(dh)
+
+        try:
+
+
+            plt.figure()
```
Comment (jklenzing) [inline pysatMagVect/tests/test_apex.py:150]: too many blank lines

File: pysatMagVect/tests/test_transformations_and_tracing.py [hunk 23]
```diff
@@ -0,0 +1,789 @@
...
+import pysat
+
+from pysatMagVect.tests.test_core import gen_data_fixed_alt, gen_trace_data_fixed_alt
+from pysatMagVect.tests.test_core import gen_plot_grid_fixed_alt
+from pysatMagVect.tests.test_core import dview, dc
+
+# results from omniweb calculator
+omni_list = [[550. , 20.00   , 0.00 , 29.77,  359.31,  -9.04 ,   3.09],
+    [550. , 20.00   , 7.50 , 29.50  ,  7.19 , -8.06 ,   9.54],
+    [550. , 20.00   ,15.00 , 28.34  , 15.01 , -7.51 ,  16.20],
```
Comment (jklenzing) [inline pysatMagVect/tests/test_transformations_and_tracing.py]: indentation

File: pysatMagVect/tests/test_transformations_and_tracing.py [hunk 24]
```diff
@@ -0,0 +1,789 @@
...
+#
+#             for i,p_lat in enumerate(p_lats):
+#                 print ('collecting 3', i, p_lat)
+#                 x2, y2, z2, _, _, h2 = pending.pop(0).get()
+#                 apex_x[i,:-1] = np.abs(x2 - x)
+#                 apex_y[i,:-1] = np.abs(y2 - y)
+#                 apex_z[i,:-1] = np.abs(z2 - z)
+#
+#
+#         else:
```
Comment (jklenzing) [inline pysatMagVect/tests/test_transformations_and_tracing.py]: Do we need commented code?

File: setup.py [hunk 25]
```diff
@@ -27,26 +23,23 @@
 
 # call setup
 #--------------------------------------------------------------------------
-numpy.distutils.core.setup( 
-# setuptools.setup( 
+numpy.distutils.core.setup(
+# setuptools.setup(
```
Comment (jklenzing) [inline setup.py]: commented code

File: setup.py [hunk 26]
```diff
@@ -1,22 +1,18 @@
-# Copyrighted by The University of Texas at Dallas subject to the United States 
+# Copyrighted by The University of Texas at Dallas subject to the United States
 # government unlimited rights (DFAR 252.227-7013-5).
 
 import setuptools
```
Comment (jklenzing) [inline setup.py:1]: package not used

File: setup.py [hunk 27]
```diff
@@ -1,22 +1,18 @@
...
-from distutils.command.sdist import sdist
-
-config = Configuration('pysatMagVect')
 
 # create extension for calling IGRF
-config.add_extension(name='igrf',
-    sources = [os.path.join('pysatMagVect', 'igrf12.f')]
-    )
+extensions = [Extension(name='pysatMagVect.igrf',
+                        sources = [os.path.join('pysatMagVect', 'igrf13.f')]),
```
Comment (jklenzing) [inline setup.py]: no whitespace around '=' for kwargs

File: setup.py [hunk 28]
```diff
@@ -1,22 +1,18 @@
...
+              Extension(name='pysatMagVect.fortran_coords',
+                        sources = [os.path.join('pysatMagVect', '_coords.f')])]
```
Comment (jklenzing) [inline setup.py]: same

File: setup.py [hunk 29]
```diff
@@ -27,26 +23,23 @@
 
 # call setup
 #--------------------------------------------------------------------------
-numpy.distutils.core.setup( 
-# setuptools.setup( 
+numpy.distutils.core.setup(
+# setuptools.setup(
     name = 'pysatMagVect',
-    version = version,        
+    version = version,
```
Comment (jklenzing) [inline setup.py]: no whitespace

**Review by jklenzing — COMMENTED** [review]
Mostly style comments caught by flake8.  I'd recommend moving plots out of unit tests and into a demo folder to speed up travis.

File: pysatMagVect/tests/test_comparison.py (hunk 1 above)
Comment (rstoneback) [reply pysatMagVect/tests/test_comparison.py:22]: Added

File: pysatMagVect/tests/test_comparison.py (hunk 2 above)
Comment (rstoneback) [reply pysatMagVect/tests/test_comparison.py:92]: For this package unit test plots are the way to go. I've used every plot to get the code to where it is at.

File: pysatMagVect/_core.py (hunk 3 above)
Comment (rstoneback) [reply pysatMagVect/_core.py:117]: I added a ecef_to_geodetic function written in Fortran that I wanted the code to use by default. I kept the python version to check accuracy and it is easier to read.

File: pysatMagVect/_core.py (hunk 8 above)
Comment (rstoneback) [reply pysatMagVect/_core.py:183]: Fixed

File: pysatMagVect/_core.py (hunk 9 above)
Comment (rstoneback) [reply pysatMagVect/_core.py]: fixed

File: pysatMagVect/_core.py (hunk 10 above)
Comment (rstoneback) [reply pysatMagVect/_core.py]: removed

File: pysatMagVect/_core.py (hunk 14 above)
Comment (rstoneback) [reply pysatMagVect/_core.py:1180]: Fixed

File: setup.py (hunk 26 above)
Comment (rstoneback) [reply setup.py:1]: Turns put python setup.py develop needs this import statement to work, even though the package isn't obviously used within setup.py

File: pysatMagVect/_core.py (hunk 19 above)
Comment (rstoneback) [reply pysatMagVect/_core.py:1704]: It is used elsewhere (step_along_..) with the unit vectors outputs assigned to variable names. More importantly, there is downstream code that depends upon pysatMagVect and changing the outputs here causes a ripple of work. Instead, I added a new method (calculate_geomagnetic_basis) with limited input options and an output dictionary that just transparently calls the calculate_mag_drift_unit_vectors. Existing code gets to remain the same and new code can use the simpler, clearer interface.

**Comment by rstoneback** [comment]
That should cover most everything:
 - Answered the questions
 - Added docstrings to test code
 - Made numerous formatting improvements
 - Moved Notes to after Return in docstrings
 - Added new method with a single output dict for the geomagnetic basis vectors. Existing method kept for compatibility of downstream code.
 - One request not currently filled. The unit tests are still making plots. The figures are a big part of testing for this package. It is also true that when I run testing locally I set it to high resolution and use the plots in the developing manuscript.

**Comment by rstoneback** [comment]
Thanks for the review @aburrell and @jklenzing . Let's talk about the plots tomorrow. Cheers.

**Comment by rstoneback** [comment]
I added issue #38 to improve the unit tests by removing the plots. I also modified manifest.in to exclude the test directory for pip produced distributions. Standard users won't see the unit tests. I plan on moving forward with the release. Thanks again for the feedback.

File: pysatMagVect/_core.py (hunk 6 above)
Comment (rstoneback) [reply pysatMagVect/_core.py:23]: Fixed.

File: .travis.yml (hunk 5 above)
Comment (rstoneback) [reply .travis.yml:7]: Added

File: pysatMagVect/tests/test_comparison.py (hunk 2 above)
Comment (rstoneback) [reply pysatMagVect/tests/test_comparison.py:92]: Added issue #38 to ensure this is resolved.

**Pull request merged**